src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1493,19 +1493,24 @@
 
 ;; Wrap the fact-type so that Clojure equality and hashcode semantics are used
 ;; even though this is placed in a Java map.
-(deftype AlphaRootsWrapper [roots roots-hash]
+(deftype AlphaRootsWrapper [fact-type fact-type-hash roots]
   Object
   (equals [this other]
-    ;; Since we memoize the function that retrieves these on the fact type, and since the alpha roots for
-    ;; any given type primary type T (excluding descendants and ancestors) are a disjoint
-    ;; set with respect to the alpha roots for any other primary type T2, we know that we will
-    ;; only have one AlphaRootsWrapper object for any given set of alpha roots per get-alphas-fn instance.
-    ;; Therefore we can just use reference equality in the interest of maximizing performance.
-    (identical? this other))
+    (let [other ^AlphaRootsWrapper other]
+      (cond
+        
+        (identical? fact-type (.fact-type other)) 
+        true
+        
+        (not (== fact-type-hash (.fact-type-hash other)))
```

Comment:
@WilliamParker thanks for adding all the details.

I should have mentioned that I too did look a bit into what happens with primitive type hints in the emitted bytecodes.  That's why I mentioned that the only threat would be that Clojure will, unlike Java, throw overflow exceptions when coercing/downcasting by default.  Which is of course changeable with the setting of `*unchecked-math*` (you called it `*unboxed-math*` above, which I think was a typo).

Just for another future note, you can always do:
```clj
(binding [*compile-files* true]
  (eval '(AlphaRootsWrapper. nil 1 nil)))
```

And look into the directory specified by the `*compile-path*` for the compiled class files.  It'll have a name derived from the current `*ns*` and probably the word "eval" in the class name.

You can use
```clj
javap -c <file name here>
```